File: 09-CustomDialog-VB/Dialogs/YouTubeVideoInsertDialog.vb
Imports System
Imports System.Windows.Forms
Imports SpiceLogic.HtmlEditor.Abstractions.Dialogs
Imports SpiceLogic.HtmlEditor.Abstractions.Entities
Imports SpiceLogic.HtmlEditor.WinForms.Helpers
Imports SpiceLogic.HtmlEditor.WinForms.Models.Dialogs

Namespace Global.CustomDialog.Dialogs

    ''' <summary>
    ''' Class YouTubeVideoInsertDialog
    ''' </summary>
    Partial Public Class YouTubeVideoInsertDialog
        Inherits DialogBase
        Implements IYouTubeVideoInsertDialog

        ''' <summary>
        ''' The _the original element
        ''' </summary>
        Private _theOriginalElement As YouTubeVideoElement

        ''' <summary>
        ''' Initializes a new instance of the <see cref="YouTubeVideoInsertDialog" /> class.
        ''' </summary>
        Public Sub New()
            InitializeComponent()

            Icon = Constants.FormIcon
        End Sub

        ''' <summary>
        ''' Gets or sets the element.
        ''' </summary>
        ''' <value>The element.</value>
        Public Property Element As YouTubeVideoElement Implements IYouTubeVideoInsertDialog.Element
            Get
                If _theOriginalElement Is Nothing Then
                    Dim theElement As YouTubeVideoElement = New YouTubeVideoElement With {
                        .Url = txtUrl.Text.Trim(),
                        .Width = txtWidth.Text,
                        .Height = txtHeight.Text
                    }
                    Return theElement
                End If
                _theOriginalElement.Url = txtUrl.Text.Trim()
                _theOriginalElement.Width = txtWidth.Text
                _theOriginalElement.Height = txtHeight.Text
                _theOriginalElement.CssStyle = txtCssStyle.Text
                Return _theOriginalElement
            End Get
            Set(value As YouTubeVideoElement)
                Me._theOriginalElement = value
                ' Spec 019 FR-009: preserve design-time defaults on empty fields.
                If Not String.IsNullOrEmpty(value.Url) Then
                    txtUrl.Text = value.Url
                End If
                If Not String.IsNullOrEmpty(value.Height) Then
                    txtHeight.Text = value.Height
                End If
                If Not String.IsNullOrEmpty(value.Width) Then
                    txtWidth.Text = value.Width
                End If
                If Not String.IsNullOrEmpty(value.CssStyle) Then
                    txtCssStyle.Text = value.CssStyle
                End If
            End Set
        End Property

        ''' <summary>
        ''' Handles the Click event of the btnOK control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub btnOK_Click(sender As Object, e As EventArgs) Handles btnOK.Click
            If String.IsNullOrEmpty(txtUrl.Text) Then
                MessageBox.Show("The YouTube URL cannot be empty.")
                txtUrl.Focus()
                Return
            End If

            Dim theUrl As String = txtUrl.Text.Trim()
            If theUrl = String.Empty Then
                MessageBox.Show("The YouTube URL cannot be empty.")
                txtUrl.Focus()
                Return
            End If

            If Not theUrl.ToLower().Contains("youtube.com") Then
                MessageBox.Show("The URL you provided does not contain the YouTube Domain name", "Invalid URL")
                Return
            End If

            Me.DialogResult = DialogResult.OK
        End Sub

        ' VB.NET requires explicit interface implementation for the IDialog
        ' members the WinForms Form base class supplies (ShowDialog from
        ' IDialog, Dispose from IDisposable). C# satisfies these implicitly.
        Public Shadows Function ShowDialog() As DialogResult Implements IDialog.ShowDialog
            Return MyBase.ShowDialog()
        End Function
    End Class

End Namespace

File: 09-CustomDialog/Dialogs/YouTubeVideoInsertDialog.cs
using System;
using System.Windows.Forms;
using SpiceLogic.HtmlEditor.Abstractions.Entities;
using SpiceLogic.HtmlEditor.WinForms.Helpers;
using SpiceLogic.HtmlEditor.WinForms.Models.Dialogs;

namespace CustomDialog.Dialogs;

/// <summary>
/// Class YouTubeVideoInsertDialog
/// </summary>
public partial class YouTubeVideoInsertDialog : DialogBase, IYouTubeVideoInsertDialog
{
    /// <summary>
    /// The _the original element
    /// </summary>
    private YouTubeVideoElement _theOriginalElement;

    /// <summary>
    /// Initializes a new instance of the <see cref="YouTubeVideoInsertDialog" /> class.
    /// </summary>
    public YouTubeVideoInsertDialog()
    {
        InitializeComponent();

        Icon = Constants.FormIcon;
    }

    /// <summary>
    /// Gets or sets the element.
    /// </summary>
    /// <value>The element.</value>
    public YouTubeVideoElement Element
    {
        get
        {
            if(_theOriginalElement == null)
            {
                YouTubeVideoElement theElement = new()
                {
                    Url = txtUrl.Text.Trim(),
                    Width = txtWidth.Text,
                    Height = txtHeight.Text
                };
                return theElement;
            }
            _theOriginalElement.Url = txtUrl.Text.Trim();
            _theOriginalElement.Width = txtWidth.Text;
            _theOriginalElement.Height = txtHeight.Text;
            _theOriginalElement.CssStyle = txtCssStyle.Text;
            return _theOriginalElement;
        }
        set {
            this._theOriginalElement = value;
            txtUrl.Text = value.Url;
            txtHeight.Text = value.Height;
            txtWidth.Text = value.Width;
            txtCssStyle.Text = value.CssStyle;
        }
    }

    /// <summary>
    /// Handles the Click event of the btnOK control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void btnOK_Click(object sender, EventArgs e)
    {
        if(string.IsNullOrEmpty(txtUrl.Text))
        {
            MessageBox.Show("The YouTube URL cannot be empty.");
            txtUrl.Focus();
            return;
        }

        string theUrl = txtUrl.Text.Trim();
        if(theUrl == string.Empty)
        {
            MessageBox.Show("The YouTube URL cannot be empty.");
            txtUrl.Focus();
            return;
        }

        if(!theUrl.ToLower().Contains("youtube.com"))
        {
            MessageBox.Show("The URL you provided does not contain the YouTube Domain name", "Invalid URL");
            return;
        }

        this.DialogResult = DialogResult.OK;
    }
}